src/dbnode/storage/index/block.go

```diff
@@ -772,23 +774,50 @@ func (b *block) Query(
 	}
 
 	size := results.Size()
-	limitedResults := false
+	batch := b.docsPool.Get()
+	batchSize := cap(batch)
+	if batchSize == 0 {
+		batchSize = defaultQueryDocsBatchSize
+	}
+	batchOpts := AddDocumentsBatchResultsOptions{
+		AllowPartialUpdates: true,
+		LimitSize:           opts.Limit,
+	}
 	iterCloser := safeCloser{closable: iter}
 	execCloser := safeCloser{closable: exec}
 
 	defer func() {
+		b.docsPool.Put(batch)
 		iterCloser.Close()
 		execCloser.Close()
 	}()
 
 	for iter.Next() {
 		if opts.LimitExceeded(size) {
-			limitedResults = true
 			break
 		}
 
-		d := iter.Current()
-		_, size, err = results.AddDocument(d)
+		batch = append(batch, iter.Current())
+		if len(batch) < batchSize {
+			continue
+		}
+
+		_, size, err = results.AddDocumentsBatch(batch, batchOpts)
+		if err != nil {
+			return false, err
+		}
+
+		// Reset batch
+		var emptyDoc doc.Document
+		for i := range batch {
+			batch[i] = emptyDoc
```

Comment:
Worth doing some sort of Memset optimization here instead?

Reply:
This is the memset optimization thankfully, just using a single zero var value and then setting everything with `range i := ... ` to that value.